src/features/code_actions.zig

```diff
@@ -166,64 +164,67 @@ fn handleUnusedVariableOrConstant(builder: *Builder, actions: *std.ArrayListUnma
-    const is_value_discarded = std.mem.eql(u8, offsets.locToSlice(builder.handle.text, capture_locs.value), "_");
-    if (is_value_discarded) {
-        // |_, i| ->
-        // TODO fix formatting
-        try actions.append(builder.arena, .{
-            .title = "remove capture",
-            .kind = .quickfix,
-            .isPreferred = true,
-            .edit = try builder.createWorkspaceEdit(&.{builder.createTextEditLoc(capture_locs.loc, "")}),
-        });
-    } else {
-        // |v, i| -> |v|
-        // |v, _| -> |v|
-        try actions.append(builder.arena, .{
-            .title = "remove index capture",
-            .kind = .quickfix,
-            .isPreferred = true,
-            .edit = try builder.createWorkspaceEdit(&.{builder.createTextEditLoc(
-                .{ .start = capture_locs.value.end, .end = capture_locs.loc.end - 1 },
-                "",
-            )}),
-        });
+    // look for next non-whitespace after last '|'. if its a '{' we can insert discards.
+    // this means bare loop/switch captures (w/out curlies) aren't supported.
+    var block_start = capture_loc.end + 1;
+    while (block_start < builder.handle.text.len and
+        std.ascii.isWhitespace(builder.handle.text[block_start])) : (block_start += 1)
+    {}
+    if (builder.handle.text[block_start] != '{') {
+        return;
     }
-}
 
-fn handleUnusedCapture(builder: *Builder, actions: *std.ArrayListUnmanaged(types.CodeAction), loc: offsets.Loc) !void {
-    const tracy_zone = tracy.trace(@src());
-    defer tracy_zone.end();
+    const block_start_loc = offsets.Loc{ .start = block_start + 1, .end = block_start + 1 };
+    const identifier_name = builder.handle.text[loc.start..loc.end];
+    const new_text = try createDiscardText(builder, identifier_name, block_start, true);
+    const action1 = .{
+        .title = "discard capture",
+        .kind = .@"source.fixAll",
+        .isPreferred = true,
+        .edit = try builder.createWorkspaceEdit(&.{builder.createTextEditLoc(block_start_loc, new_text)}),
+    };
+    const action2 = .{
+        .title = "discard capture name",
+        .kind = .quickfix,
+        .isPreferred = false,
+        .edit = try builder.createWorkspaceEdit(&.{builder.createTextEditLoc(loc, "_")}),
+    };
 
-    const capture_locs = getCaptureLoc(builder.handle.text, loc, false) orelse return;
+    // prevent adding duplicate 'remove capture' action.
+    // search for a matching action by comparing ranges.
+    const remove_cap_loc = builder.createTextEditLoc(capture_loc, "");
+    const found_remove_capture_action = outer: for (actions.items) |item| {
+        const edit = item.edit orelse continue;
+        const changes = edit.changes orelse continue;
+        var iter = changes.iterator();
+        while (iter.next()) |entry| {
+            const t_edits = entry.value_ptr.*;
+            for (t_edits) |t_edit| {
+                if (remove_cap_loc.range.start.line == t_edit.range.start.line and
```

Comment:
At a glance this seems HashMapable, I like the idea you mentioned where we'd pass around the map with actions :)